src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -3299,6 +3301,277 @@ tx_gateway_frontend::get_all_transactions() {
+                if (!r2) {
+                    vlog(
+                      txlog.info, "got {} on updating fetched tx", r2.error());
+                    co_return tx_errc::not_coordinator;
+                }
+                tx = r2.value();
+            } else {
+                vlog(
+                  txlog.warn,
+                  "fetched a tx with unexpected tx_seq:{}",
+                  old_tx.tx_seq);
+                co_return tx_errc::unknown_server_error;
+            }
+        } else if (
+          old_tx.status == tm_transaction::tx_status::prepared
+          || old_tx.status == tm_transaction::tx_status::aborting) {
+            if (old_tx.tx_seq != tx.tx_seq || old_tx.status != tx.status) {
+                vlog(
+                  txlog.warn,
+                  "Fetched tx and current do not mutch. Fetched: status:{}, "
+                  "tx_seq:{}. Current: status:{}, tx_seq:{}",
+                  old_tx.get_status(),
+                  old_tx.tx_seq,
+                  tx.get_status(),
+                  tx.tx_seq);
+                co_return tx_errc::unknown_server_error;
+            }
+            old_tx.etag = term;
+            auto r2 = co_await stm->update_tx(old_tx, term);
+            if (!r2) {
+                vlog(txlog.info, "got {} on updating fetched tx", r2.error());
+                co_return tx_errc::not_coordinator;
+            }
+            tx = r2.value();
+        } else {
+            vlog(
+              txlog.warn,
+              "fetched a tx with an unexpected status:{}",
+              old_tx.get_status());
+            co_return tx_errc::unknown_server_error;
+        }
+    } else if (tx.status == tm_transaction::tx_status::ongoing) {
+        auto r1 = co_await fetch_tx(tx.id, tx.etag);
+        if (!r1) {
+            vlog(
+              txlog.warn,
+              "can't find old tx:{} etag:{} during describe_tx call",
+              tx.id,
+              tx.etag);
+            // It means we should kill current tx and return error
+            co_await stm->expire_tx(tx.id);
+            co_return tx_errc::invalid_txn_state;
+        }
+
+        auto old_tx = r1.value();
+
+        if (old_tx.pid != tx.pid) {
+            vlog(
+              txlog.warn,
+              "fetched pid:{} doesn't match local tx's pid:{}",
+              old_tx.pid,
+              tx.pid);
+            co_return tx_errc::unknown_server_error;
+        }
+
+        if (old_tx.status != tm_transaction::tx_status::ongoing) {
+            vlog(
+              txlog.warn,
+              "fetched a tx with an unexpected status:{}",
+              old_tx.status);
+            co_return tx_errc::unknown_server_error;
+        }
+
+        old_tx.etag = term;
+        auto r2 = co_await stm->update_tx(old_tx, term);
+        if (!r2) {
+            vlog(txlog.info, "got {} on updating fetched tx", r2.error());
+            co_return tx_errc::not_coordinator;
+        }
```

Comment:
pull the update_tx from all the branches, let them fall through and update the etag just before `co_return tx;` in this method, otherwise there is a lot of duplication; just make sure that we update `tx = old_tx` when `old_tx` is fresher